Mapping immutable types #216

piotrdytkowski opened the pull request:
Allows to use no-args constructor to create immutable objects

garethahealy commented:
@piotrdytkowski ; can you rebase with latest to force travis to retest

garethahealy commented:
@piotrdytkowski ; any update?

piotrdytkowski commented:
Hey, sure, will have a look later. I don't like the name 'omit-constructor' anymore though...

garethahealy commented:
@piotrdytkowski ; have you had anytime to progress?

piotrdytkowski commented:
Hey, I rebased the project, but now have a test failing in maven but passing in my IDE, must be something with deps or environment. I hope to sort it out soon.

piotrdytkowski commented:
Now I see that master branch ci is failing for the same reason. I will update when you fix/revert master.

garethahealy commented:
@piotrdytkowski ; master branch has now been fixed. Can you rebase.

piotrdytkowski commented:
I rebased but got some osgi tests failing, they seem unrelated but I can be wrong. Can you help me with this?
```
Tests in error: 
  OsgiContainerTest.shouldInstantiate:59 Â» ClassNotFound org.dozer.DozerBeanMapp...
  OsgiContainerTest.shouldLoadMappingFile:72 Â» ClassNotFound org.dozer.DozerBean...
  OsgiContainerTest.shouldMap:65 Â» ClassNotFound org.dozer.DozerBeanMapperBuilde...
```

piotrdytkowski commented:
Hey, did you have any time to look at this?

piotrdytkowski commented:
Hey @garethahealy , did you have any time to look at this?

garethahealy commented:
@piotrdytkowski ; no sorry, am currently on holiday, so might be a few weeks.

garethahealy commented on core/pom.xml:
```diff
@@ -115,12 +115,30 @@
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+            <version>4.12</version>
+        </dependency>
```
junit isnt required. It comes from the parent pom, as per:
- https://github.com/DozerMapper/dozer/blob/master/plugins-parent/pom.xml#L83-L87

garethahealy commented on core/pom.xml:
```diff
@@ -115,12 +115,30 @@
+            <groupId>org.apache.maven.surefire</groupId>
+            <artifactId>surefire-junit4</artifactId>
+            <version>2.20</version>
+        </dependency>
```
What uses this dep? Cant see why its needed.

garethahealy commented on schema/pom.xml:
```diff
@@ -43,6 +43,19 @@
+            <groupId>org.apache.maven.surefire</groupId>
+            <artifactId>surefire-junit4</artifactId>
+            <version>2.20</version>
+        </dependency>
```
Same as other comment. Why are these 2 needed?

garethahealy commented:
@piotrdytkowski ; the OSGi test are failing as you've added a new dependancy: org.objenesis.

It needs to be added as a required dependancy in the OSGi tests:
- https://github.com/DozerMapper/dozer/blob/master/osgi-test/src/test/java/com/github/dozermapper/osgitests/OsgiContainerTest.java#L56-L61

piotrdytkowski commented:
dozer-osgi-tests has a dependency on dozer-core so it should pull objenesis just fine. I rebased again and removed test libs added before. Hope to merge it soon :champagne:

piotrdytkowski commented:
The problem is that maven build fails for me even without my changes (errors in osgi test) so I can't make any progress. I would need some help with the project setup. Otherwise it won't be merged before 2025.

garethahealy commented:
Like i said, you need to add the new dep to the OSGi container. Since my last comment, this has slightly changed, but if you look at:
- https://github.com/DozerMapper/dozer/blob/master/tests/dozer-osgi-tests/src/test/java/com/github/dozermapper/osgitests/karaf/BundleOptions.java#L26

You'll see all the core deps needed. Those "link" URLS come from build time in:
- target/generated-test-resources/alta/

piotrdytkowski commented:
Great, that's what I needed. Thanks :)

piotrdytkowski commented:
Build is green.

dmitrySDK commented:
Hi! I think it is great functionality, but I cannot find a way to configure skipConstructor without XML, that is only with API. Is that possible with current code or maybe it should be a different PR?

BR
Dmitry Sherstnyakov

piotrdytkowski commented:
I didn't have a look at the api, back in 2014 when I implemented it maybe api configuration didn't even exist :D
I would suggest to get it merged and then build on top of it. Maybe I'll do it when I have some free time.

garethahealy commented on core/pom.xml:
```diff
@@ -161,6 +161,13 @@
+        <dependency>
+            <groupId>org.objenesis</groupId>
+            <artifactId>objenesis</artifactId>
+            <version>2.6</version>
```
Can you add this to the dependancyManagement section of the bom
- https://github.com/DozerMapper/dozer/blob/master/bom-dependencies/pom.xml

garethahealy commented:
CC @orange-buffalo

garethahealy commented:
Looks fine from my point of view.

piotrdytkowski commented:
Review the changes, please. I had to make that last commit to retrigger the build, because it ... randomly failed. Thanks for the feedback, good spot with the ObjectInstantiator. Thanks :+1:
